### PR TITLE
📖  Start updating controllers in architecture doc

### PR DESCRIPTION
**Pull request by MikeSpreitzer**
## Summary
This PR makes some updates to the architecture document, to correct old errors and track work that has been done over the past year.

Preview at https://mikespreitzer.github.io/kcp-edge-mc/doc-update-1028/direct/architecture/

## Related issue(s)

Fixes #

### CODE REVIEW

File: docs/content/direct/architecture.md [hunk 1]
```diff
@@ -77,13 +82,37 @@ More details on the internals of this module are provided in [KubeStellar Contro
...
-* The pluggable transport controller watches `Binding` objects on the WDS, and maintains in the Inventory and Transport Space (ITS) 
-a wrapped object per `Binding` to be delivered.
-* This controller is pluggable and can potentially be implemented using different options. Currently the only option we support is based on the [Open Cluster Management Project](https://open-cluster-management.io) 
-
-There is one instance of the pluggable transport controller for each WDS. 
-Currently this controller runs in an executable process. This is a work in progress and we're working on running this controller in a dedicated pod.
-More details on the internals of this module are provided in [KubeStellar Controllers Architecture](#kubestellar-controllers-architecture).
+This controller's job is to (a) get workload objects from WDS to WECs
+as prescribed by the `Binding` objects and their referenced
+`CustomTransform` objects and inventory objects and (b) get
```
Comment (waltforme) [inline docs/content/direct/architecture.md:87]: Is (b) done by the OCM status addon?

File: docs/content/direct/architecture.md [hunk 2]
```diff
@@ -77,13 +82,37 @@ More details on the internals of this module are provided in [KubeStellar Contro
...
+corresponding reported state back into `WorkStatus` objects in the
+ITS.
+
+Different implementations of this controller are possible; it would be
+possible to enable even more different implementations by taking a
+more general approach to inventory.
+
+The implementations need not be in this Git repository. Currently
+there is one iimplementation, and it _is_ in this repository. This
```
Comment (waltforme) [inline docs/content/direct/architecture.md:96]: s/iimplementation/implementation/

File: docs/content/direct/architecture.md [hunk 3]
```diff
@@ -201,55 +230,89 @@ events, work queues for parallel processing of tasks, and a reconciler
...
-The Binding controller is responsible for watching workload objects and 
-`BindingPolicy` objects, and maintains for each of the latter a matching `Binding` object in the WDS. 
-A `Binding` object is mapped 1:1 to a `BindingPolicy` object and contains references to the concrete list of workload 
-objects and references to the concrete list of destinations that were selected by the policy.
-
+The Binding controller is responsible for watching workload objects
+and `BindingPolicy` objects, and maintains for each of the latter a
+matching `Binding` object in the WDS.  A `Binding` object is mapped
+1:1 to a `BindingPolicy` object and contains the concrete list of
+references workload objects and the concrete list of references to
```
Comment (waltforme) [inline docs/content/direct/architecture.md:289]: references _to_ (workload objects)?

File: docs/content/direct/architecture.md (hunk 1 above)
Comment (MikeSpreitzer) [reply docs/content/direct/architecture.md:87]: This paragraph is stating the responsibilities of the chosen transport; how the one based on OCM gets these done comes later.

Comment (MikeSpreitzer) [reply docs/content/direct/architecture.md:87]: There is a bigger conceptual/terminological muddle that already existed and I am _not_ trying to untangle in this PR. I would say that KubeStellar has "pluggable transport", and the one implementation is based on OCM and involves several things: the ocm-transport-controller Deployment/Pod, the use of OCM, and the extension of OCM with the OCM Status Add-On (which, like every OCM extension, has both a "Controller" and an "Agent").

File: docs/content/direct/architecture.md (hunk 3 above)
Comment (MikeSpreitzer) [reply docs/content/direct/architecture.md]: fixed

File: docs/content/direct/architecture.md (hunk 2 above)
Comment (MikeSpreitzer) [reply docs/content/direct/architecture.md]: fixed.

File: docs/content/direct/architecture.md (hunk 1 above)
Comment (MikeSpreitzer) [reply docs/content/direct/architecture.md:88]: I have added a few more words, trying to be a little clearer without taking on a major rewriting of the whole topic.

**Comment by MikeSpreitzer** [comment]
The force-push to e6ff37de8 is for edits responding to the review comments above.

File: docs/content/direct/architecture.md [hunk 4]
```diff
@@ -61,14 +61,19 @@ and updates `WorkStatus` objects in the ITS namespace associated with the WEC.
...
+This module manages the binding controller and the status controller. 
 
-* The binding controller watches `BindingPolicy` and workload objects on the Workload Definition Space (WDS), and maintains
-a `Binding` object for each `BindingPolicy` in the WDS. The `Binding` object contains references to the concrete list of
-workload objects and references to the concrete list of clusters that were selected by the `BindingPolicy` selectors.
+* The binding controller watches `BindingPolicy` and workload objects
+on the Workload Definition Space (WDS), and maintains a `Binding`
+object for each `BindingPolicy` in the WDS. A `Binding` object
+contains (a) the concrete list of references to workload objects (and
+associated modulations on downsync behavior) and (b) the concrete list
```
Comment (vMaroon) [inline docs/content/direct/architecture.md:70]: Perhaps should add a hyperlink to the mentioned modulations on downsync behavior.

File: docs/content/direct/architecture.md [hunk 5]
```diff
@@ -201,55 +231,89 @@ events, work queues for parallel processing of tasks, and a reconciler
...
+internal data structure, and the ability to read from it. The other is
+notifying about workload object events.
 
 The architecture and the event flow of the code for create/update object events is
-illustrated in Figure 3 (some details might be omitted to make the flow easier
-to understand). 
+illustrated in Figure 3 (some details are omitted to make the flow easier
+to understand).
 
 ![Figure 3 - Binding Controller](./images/binding-controller.svg)
```
Comment (vMaroon) [inline docs/content/direct/architecture.md:318]: Note for later: this diagram is not dark-mode friendly.

File: docs/content/direct/architecture.md [hunk 6]
```diff
@@ -263,166 +327,98 @@ handler and the work queue as follows:
     - Creates GVR key
     - Registers Event Handler
     - Starts Informer
-    - Indexes informer and lister in a map by GVR key
+    - Stores informer and lister in a map indexed by GVR
 - Waits for all caches to sync
 - Starts N workers to process work queue
```
Comment (vMaroon) [inline docs/content/direct/architecture.md:333]: Missing from docs: at startup the binding controller currently notes all binding-policies.

File: docs/content/direct/architecture.md [hunk 7]
```diff
@@ -263,166 +327,98 @@ handler and the work queue as follows:
...
 
-- Flow continues the same way as in the create/update scenario, however
-  the `deletedObject` field in the key indicates that the object has been
-  deleted and that it needs to be removed from all clusters.
+1. Ensure the absence of the KubeStellar finalizer on the `BindingPolicy`.
+1. Invoke the `BindingPolicyResover` method for the absence of the `BindingPolicy`.
 
-- In-memory representations of affected bindings are updated to remove the object.
-If any are indeed affected, they are enqueued for syncing.
+For a `BindingPolicy` that is not deleted or being deleted, sync consists of the following steps.
```
Comment (vMaroon) [inline docs/content/direct/architecture.md:347]: For accuracy this should say "is not deleted and is not being deleted".

File: docs/content/direct/architecture.md (hunk 4 above)
Comment (MikeSpreitzer) [reply docs/content/direct/architecture.md:70]: I am not aware of a good place to link to. #2463 should make this easier. Right now, I _could_ list them three inline here. But I think that is getting into more detail than is appropriate here (a section about the controller-manager).

File: docs/content/direct/architecture.md (hunk 6 above)
Comment (MikeSpreitzer) [reply docs/content/direct/architecture.md:334]: Revised.

File: docs/content/direct/architecture.md (hunk 7 above)
Comment (MikeSpreitzer) [reply docs/content/direct/architecture.md]: Revised.

**Comment by MikeSpreitzer** [comment]
The force-push to f4f41073a makes a few edits in response to the review above.

**Comment by MikeSpreitzer** [comment]
.. and the force-push to 560467324 adds one paragraph break in https://mikespreitzer.github.io/kcp-edge-mc/doc-update-1028/direct/architecture/#sync-workload-object

**Pull request merged**